Actions

dave-tucker opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/linuxkit/linuxkit/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

Per the discussion in #3437, having a working CI system is important to attract contributors to the project and reduce the burden on maintainers. I have spent some time looking at consolidating our CI system sprawl (CircleCI, LinuxKitCI etc..) and replacing it with GitHub Actions.

This PR introduced a single worflow for CI.
It:
- Builds binaries and runs unit tests on all supported platforms
- Runs all integration tests (in parallel)

In future this can be extended to include anything and everything that is on the CI wishlist.

A few points of note:

1. Tests are run using containerized qemu (no nested virt)
2. Overall CI time is still 1hr, but that is mainly spent running the packages tests
3. The GCP test that was running on LinuxKit CI has been ported to RTF but isn't running just yet due to security concerns around secrets
4. I'm not sure what the security model is around changes to workflow/actions. I'll clarify before merge.

**- How I did it**

Wrote lots of YAML and tested on my LinuxKit fork.

**- How to verify it**
You can see the latest green run on my fork: https://github.com/dave-tucker/linuxkit/commit/71a95bff490e7d4505767dbf337b769c58e46f16/checks?check_suite_id=329481528

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/3781777/69637766-396c0f80-1051-11ea-846a-83d602d2d881.png)

deitch commented on test/Makefile:
```diff
@@ -42,8 +42,8 @@ ltp: $(LINUXKIT) test-ltp.img.tar.gz
+test:
 	@rtf -l build -v run -x
 
-test-pr: gcp-hack
```
With `gcp-hack` no longer a dependency, can we get rid of the `gcp-hack` structure? It is [here](https://github.com/linuxkit/linuxkit/blob/71a95bff490e7d4505767dbf337b769c58e46f16/test/Makefile#L21-L30)

dave-tucker replied:
Not quite yet. It's still being used by LinuxKitCI and deleting it now would make this PR red 😉 
If we merge this, I can come in and clean up afterwards.
Not only can `gcp-hack` disappear but so can the `ci`, `ci-pr` `ci-tag` targets in the top-level file.

I also look at moving the LTP tests to RTF outside of this PR so things will be super clean

deitch replied:
Ah, I should have realized that. Of course it makes sense.

deitch commented on src/cmd/linuxkit/run_qemu.go:
```diff
@@ -452,8 +453,13 @@ func runQemuContainer(config QemuConfig) error {
+	userString := fmt.Sprintf("%s:%s", currentUser.Uid, currentUser.Gid)
 	dockerArgs := append([]string{"run", "--interactive", "--rm", "-w", cwd}, binds...)
-	dockerArgsImg := append([]string{"run", "--rm", "-w", cwd}, binds...)
+	dockerArgsImg := append([]string{"run", "--user", userString, "--rm", "-w", cwd}, binds...)
```
I assume this is necessary when running on GitHub actions. Are there any knockdown effects? This does get called anytime someone calls `lkt run ` with qemu, does it not?

dave-tucker replied:
It's not just for actions (see commit message for more info).
TL;DR on Linux a disk created using containerized `qemu-img` will be owned by `root:root`.

Adding `sudo` to a test seemed like a bad idea and this was the best alternative.

If someone with a mac (without qemu installed) could run this from  the `tests` directory it should confirm whether or not this breaks anything.
```
rtf -vvv -l build run -x linuxkit.package.extend.ext4
```

deitch replied:
I would do it, but definitely have `qemu` installed. It will not help.

justincormack replied:
I think the containerized qemu is a mistake, and we should remove support for it.

dave-tucker replied:
@justincormack I'll take a look at removing it + then we can rebase this PR on top of that.

dave-tucker commented:
Additional note to reviewers. I've got a help request in to GitHub support to ask:

1) What happens if contributors open PRs against the workflow?
2) Whether self-hosted runners that are only used for `push` events (maintainers have already reviewed code) would prevent forks from gaining access to run untrusted code.

I'll update once I hear back.

dave-tucker commented:
Added `[DO NOT MERGE]` here since this now:

a) Depends on #3443 
b) Needs a little more testing since I added an action for `rtf` and attempted to use caching of the binary to shave a few minutes off the overall build time.

dave-tucker commented:
Rebased since the QEMU patch has merged.
Ready for review.
Still waiting to hear back from GitHub support though 😢

justincormack commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,322 @@
+  
+    - name: Symlink Linuxkit
+      run: |
+        sudo ln -s `pwd`/lkt/linuxkit-amd64-linux /bin/linuxkit
```
I think `/usr/local/bin/linuxkit` is better - adding stuff in `/bin` is a bit weird...

justincormack commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,322 @@
+    
+    - name: Symlink RTF
+      run: |
+        sudo ln -s `pwd`/bin/rtf /bin/rtf
```
same here

dave-tucker commented:
Comments addressed @justincormack!
I think this in a state where it can be merged and safely replace CircleCI and LinuxKitCI.

The only delta would be the `lkt run gcp` test that LinuxKitCI does, but that can be added back in the next piece of work I have planned.

That would be to add workflows that cover:
- Building and signing images
- Running the LTP tests
- Running integration tests for all `lkt run` platforms etc...

The approach I have in mind is to keep the logic for ☝️ in `linuxkit/actions`.
The workflows would be in `linuxkit/private-workflows` - although I need to look at how best to handle triggering.
This will keep the most sensitive parts of CI/CD ring fenced off to avoid leaking secrets or providing broad access to self-hosted runners.

deitch commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,322 @@
+    strategy:
+      matrix:
+        arch:
+          - amd64-linux
```
I do wish that there were a way to source this from some other file, so that the "official supported matrix" would be available as a core part of the repo at the top level, and not buried deep under `.github/...`. Is that doable? Or not really?

dave-tucker replied:
I don't think it's do-able as actions doesn't support YAML anchors or anything like that.

deitch replied:
Looks like it. Oh well, it was a nice try.

deitch commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,322 @@
+      uses: actions/setup-go@v1
+      with:
+        go-version: 1.11
+      id: go
```
Pretty old version, is it not?

dave-tucker replied:
Yes it is, but it's what's currently being used in CI. I didn't want to bump it here.

deitch replied:
That is a very good point. I think I will open an issue separately.

justincormack commented:
Lets go ahead and then iterate on this.

rn commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,322 @@
+
+    - name: Build Packages
+      run: |
+        make -C pkg build
```
does ubuntu-latest have docker pre-installed? If so which version? Is there a way to pin the version?

dave-tucker replied:
According to [Github](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/software-installed-on-github-hosted-runners) we have Docker version 3.0.8 🤔 
Sadly there is no way I know of to pin this.
I'd like very much to remove the Docker dependency anyway

rn commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,322 @@
+
+    - name: Symlink Linuxkit
+      run: |
+        sudo ln -s `pwd`/lkt/linuxkit-amd64-linux /bin/linuxkit
```
Is there a reason we link `linuxkit` to `/bin/linuxkit` but `rtf` to `/usr/local/bin/rtf`? Would be good to add a comment to the relevant commit if there is a good reason.

dave-tucker replied:
No reason. I can fix this quickly before we merge.

rn replied:
SGTM

rn commented:
I noticed other PRs also failing on the GCP error. Not sure if anything changed behind the scenes or if this is a regression

dave-tucker commented:
Based on the error message it looks like it's something behind the scenes. The Python script that allocates machines in GCP gets a HTTP 503 when trying to list instances.
